Google.PowerShell.IntegrationTests/Storage/Storage.GcsAcl.Tests.ps1

```diff
@@ -0,0 +1,585 @@
+
+        $projectViewerAcl = $bucketAcls | Where-Object {$_.Entity -like "project-viewers*"} | Select -First 1
+        $projectViewerAcl | Should Not BeNullOrEmpty
+        $projectViewerAcl.Entity | Should Match "project-viewers"
+        $projectViewerAcl.Kind | Should BeExactly $bucketAclKind
+        $projectViewerAcl.Role | Should BeExactly "READER"
+    }
+
+    It "should give access errors as appropriate" {
+        { Get-GcsBucketAcl -Name "asdf" } | Should Throw "403"
+    }
+}
+
+Describe "Add-GcsBucketAcl" {
+    $r = Get-Random
+    $script:bucketName = "gcps-testing-acl-bucket-$r"
+    New-GcsBucket -Name $bucketName
+
+    AfterAll {
+        Remove-GcsBucket -Name $script:bucketName -Force
+    }
+
+    It "should throw error for non-existent bucket" {
+        { Add-GcsBucketAcl -Name "gcps-bucket-no-exist" -Role Reader -AllUsers } | Should Throw "404"
+    }
+
+    It "should work for -AllUsers" {
+        $addedAcl = Add-GcsBucketAcl -Name $bucketName -Role Reader -AllUsers
+        $acl = Get-GcsBucketAcl -Name $bucketName | Where-Object {$_.Entity -match "allUsers"} | Select -First 1
+        CompareAcl $addedAcl $acl | Should Be $true
+
+        # Test that the newly created object doesn't have the ACL we just add.
+        $objectAcl = (New-GcsObject -Bucket $bucketName -ObjectName "test-object-$r" -Value "blah" -Force).Acl
+        ($objectAcl | Where-Object {$_.Entity -match "allUsers"}) | Should BeNullOrEmpty
+    }
+
+    It "should work for -AllAuthenticatedUsers" {
+        $addedAcl = Add-GcsBucketAcl -Name $bucketName -Role Reader -AllAuthenticatedUsers
+        $acl = Get-GcsBucketAcl -Name $bucketName | Where-Object {$_.Entity -match "allAuthenticatedUsers"} | Select -First 1
+        CompareAcl $addedAcl $acl | Should Be $true
+        # Test that the newly created object doesn't have the ACL we just add.
+        $objectAcl = (New-GcsObject -Bucket $bucketName -ObjectName "test-object-$r" -Value "blah" -Force).Acl
+        ($objectAcl | Where-Object {$_.Entity -match "allAuthenticatedUsers"}) | Should BeNullOrEmpty
+    }
+
+    It "should throw error for wrong user" {
+        { Add-GcsBucketAcl -Name $bucketName -Role Reader -User blahblahblah@example.com } |
+            Should Throw "Unknown user email address"
+    }
+
+    It "should work for valid user" {
+        $addedAcl = Add-GcsBucketAcl -Name $bucketName -Role Owner -User $userEmail
+        $acl = Get-GcsBucketAcl -Name $bucketName | Where-Object {$_.Entity -match "user-$userEmail"} | Select -First 1
+        CompareAcl $addedAcl $acl | Should Be $true
+
+        # Test that the newly created object doesn't have the ACL we just add.
+        $objectAcl = (New-GcsObject -Bucket $bucketName -ObjectName "test-object-$r" -Value "blah" -Force).Acl
+        ($objectAcl | Where-Object {$_.Entity -match "user-$userEmail"}) | Should BeNullOrEmpty
+    }
+
+    It "should throw error for wrong group" {
```

Comment:
It will not throw. It will just override the existing ACL (so in this case, nothing is changed).